client/eth: checkTxStatus returns a receipt on error sometimes.

Hit a panic while testing:
```
--- FAIL: TestContract (985.33s)
    --- PASS: TestContract/testSwap (2.10s)
    --- PASS: TestContract/testInitiate (102.21s)
    --- FAIL: TestContract/testRedeem (881.02s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x180 pc=0xc172a0]

goroutine 1973 [running]:
testing.tRunner.func1.2({0xcf6b20, 0x15662d0})
	/usr/lib/go/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
	/usr/lib/go/src/testing/testing.go:1392 +0x39f
panic({0xcf6b20, 0x15662d0})
	/usr/lib/go/src/runtime/panic.go:838 +0x207
decred.org/dcrdex/client/asset/eth.testRedeem(0xc02e17ed00)
	/home/joe/git/dcrdex/client/asset/eth/nodeclient_harness_test.go:1214 +0x22a0
testing.tRunner(0xc02e17ed00, 0xf8f098)
	/usr/lib/go/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
	/usr/lib/go/src/testing/testing.go:1486 +0x35f
exit status 2
FAIL	decred.org/dcrdex/client/asset/eth	1069.039s
```

And here is this area:
https://github.com/decred/dcrdex/blob/11fe415b1218463254f5e9ee2cdce6e2c4480416/client/asset/eth/nodeclient_harness_test.go#L998-L1004

So, the method errored, but it was not the error that returns a receipt, then we tried to get the used gas which panic'd. I think the design of `checkTxStatus` needs to be changed to take a receipt and return only an error to avoid pitfalls like what was seen in the test.

--- a/client/asset/eth/eth.go
+++ b/client/asset/eth/eth.go
@@ -2663,6 +2663,19 @@ func (w *assetWallet) isRefundable(secretHash [32]byte, contractVer uint32) (ref
 	})
 }
 
+func checkTxStatus(receipt *types.Receipt, gasLimit uint64) error {
+	if receipt.Status != types.ReceiptStatusSuccessful {
+		return fmt.Errorf("transaction status failed")
+
+	}
+
+	if receipt.GasUsed > gasLimit {
+		return fmt.Errorf("gas used, %d, appears to have exceeded limit of %d", receipt.GasUsed, gasLimit)
+	}
+
+	return nil
+}
+
 // getGasEstimate is used to get a gas table for an asset's contract(s). The
 // provided gases, g, should be generous estimates of what the gas might be.
 // Errors are thrown if the provided estimates are too small by more than a
@@ -2779,23 +2792,32 @@ func GetGasEstimates(ctx context.Context, cl *nodeClient, c contractor, maxSwaps
 			return fmt.Errorf("initiate error for %d swaps: %v", n, err)
 		}
 		waitForMined()
-		if r, err := cl.checkTxStatus(ctx, tx, txOpts); err != nil {
+		receipt, err := cl.transactionReceipt(ctx, tx.Hash())
+		if err != nil {
 			return err
-		} else {
-			stats.swaps = append(stats.swaps, r.GasUsed)
 		}
+		if err := checkTxStatus(receipt, txOpts.GasLimit); err != nil {
+			return err
+		}
+		stats.swaps = append(stats.swaps, receipt.GasUsed)
 
 		if isToken {
-			if r, err := cl.checkTxStatus(ctx, approveTx, txOpts); err != nil {
+			receipt, err = cl.transactionReceipt(ctx, approveTx.Hash())
+			if err != nil {
 				return err
-			} else {
-				stats.approves = append(stats.approves, r.GasUsed)
 			}
-			if r, err := cl.checkTxStatus(ctx, transferTx, txOpts); err != nil {
+			if err := checkTxStatus(receipt, txOpts.GasLimit); err != nil {
 				return err
-			} else {
-				stats.transfers = append(stats.transfers, r.GasUsed)
 			}
+			stats.approves = append(stats.approves, receipt.GasUsed)
+			receipt, err = cl.transactionReceipt(ctx, transferTx.Hash())
+			if err != nil {
+				return err
+			}
+			if err := checkTxStatus(receipt, txOpts.GasLimit); err != nil {
+				return err
+			}
+			stats.transfers = append(stats.transfers, receipt.GasUsed)
 		}
 
 		// Estimate a refund
@@ -2823,11 +2845,14 @@ func GetGasEstimates(ctx context.Context, cl *nodeClient, c contractor, maxSwaps
 			return fmt.Errorf("redeem error for %d swaps: %v", n, err)
 		}
 		waitForMined()
-		if r, err := cl.checkTxStatus(ctx, tx, txOpts); err != nil {
+		receipt, err = cl.transactionReceipt(ctx, tx.Hash())
+		if err != nil {
 			return err
-		} else {
-			stats.redeems = append(stats.redeems, r.GasUsed)
 		}
+		if err := checkTxStatus(receipt, txOpts.GasLimit); err != nil {
+			return err
+		}
+		stats.redeems = append(stats.redeems, receipt.GasUsed)
 	}
 	return nil
 }

--- a/client/asset/eth/nodeclient.go
+++ b/client/asset/eth/nodeclient.go
@@ -380,25 +380,6 @@ func (n *nodeClient) sendSignedTransaction(ctx context.Context, tx *types.Transa
 	return n.leth.ApiBackend.SendTx(ctx, tx)
 }
 
-func (n *nodeClient) checkTxStatus(ctx context.Context, tx *types.Transaction, txOpts *bind.TransactOpts) (*types.Receipt, error) {
-	// It appears the receipt is only accessible after the tx is mined.
-	receipt, err := n.transactionReceipt(ctx, tx.Hash())
-	if err != nil {
-		return nil, fmt.Errorf("error getting transaction receipt: %v", err)
-	}
-
-	if receipt.Status != types.ReceiptStatusSuccessful {
-		return receipt, fmt.Errorf("transaction status failed")
-
-	}
-
-	if receipt.GasUsed > txOpts.GasLimit {
-		return receipt, fmt.Errorf("gas used, %d, appears to have exceeded limit of %d", receipt.GasUsed, txOpts.GasLimit)
-	}
-
-	return receipt, nil
-}
-
 // newTxOpts is a constructor for a TransactOpts.
 func newTxOpts(ctx context.Context, from common.Address, val, maxGas uint64, maxFeeRate, gasTipCap *big.Int) *bind.TransactOpts {
 	if gasTipCap.Cmp(maxFeeRate) > 0 {

--- a/client/asset/eth/nodeclient_harness_test.go
+++ b/client/asset/eth/nodeclient_harness_test.go
@@ -786,15 +786,17 @@ func testInitiate(t *testing.T, assetID uint32) {
 			t.Fatalf("%s: post-initiate mining error: %v", test.name, err)
 		}
 
-		receipt, err := ethClient.checkTxStatus(ctx, tx, txOpts)
-		if err != nil && test.success {
-			spew.Dump(tx)
-			spew.Dump(receipt)
-			t.Fatalf("%s: failed transaction status: %v", test.name, err)
-		} else if receipt.GasUsed > expGas {
-			t.Fatalf("%s: gas used %d exceeds expected max %d", test.name, receipt.GasUsed, expGas)
+		// It appears the receipt is only accessible after the tx is mined.
+		receipt, err := ethClient.transactionReceipt(ctx, tx.Hash())
+		if err != nil {
+			t.Fatalf("%s: failed retrieving initiate receipt: %v", test.name, err)
 		}
+		spew.Dump(receipt)
 
+		err = checkTxStatus(receipt, txOpts.GasLimit)
+		if err != nil && test.success {
+			t.Fatalf("%s: failed init transaction status: %v", test.name, err)
+		}
 		fmt.Printf("Gas used for %d initiations, success = %t: %d (expected max %d) \n",
 			len(test.swaps), test.success, receipt.GasUsed, expGas)
 
@@ -896,9 +898,15 @@ func testRedeemGas(t *testing.T, assetID uint32) {
 	if err := waitForMined(t, time.Second*8, true); err != nil {
 		t.Fatalf("unexpected error while waiting to mine: %v", err)
 	}
-	_, err = ethClient.checkTxStatus(ctx, tx, txOpts)
+	receipt, err := ethClient.transactionReceipt(ctx, tx.Hash())
 	if err != nil {
-		t.Fatalf("Init transaction failed: %v", err)
+		t.Fatalf("failed retrieving initiate receipt: %v", err)
+	}
+	spew.Dump(receipt)
+
+	err = checkTxStatus(receipt, txOpts.GasLimit)
+	if err != nil {
+		t.Fatalf("failed init transaction status: %v", err)
 	}
 
 	// Make sure swaps were properly initiated
@@ -1103,11 +1111,16 @@ func testRedeem(t *testing.T, assetID uint32) {
 			t.Fatalf("%s: post-init mining error: %v", test.name, err)
 		}
 
-		receipt, err := test.redeemerClient.checkTxStatus(ctx, tx, txOpts)
+		receipt, err := test.redeemerClient.transactionReceipt(ctx, tx.Hash())
+		if err != nil {
+			t.Fatalf("%s: failed to get init receipt: %v", test.name, err)
+		}
+		spew.Dump(receipt)
+
+		err = checkTxStatus(receipt, txOpts.GasLimit)
 		if err != nil {
 			t.Fatalf("%s: failed init transaction status: %v", test.name, err)
 		}
-		spew.Dump(receipt)
 		fmt.Printf("Gas used for %d inits: %d \n", len(test.swaps), receipt.GasUsed)
 
 		for i := range test.swaps {
@@ -1161,14 +1174,17 @@ func testRedeem(t *testing.T, assetID uint32) {
 			t.Fatalf("%s: post-redeem mining error: %v", test.name, err)
 		}
 
-		expSuccess := !test.expectRedeemErr && test.addAmt
-		receipt, err = test.redeemerClient.checkTxStatus(ctx, tx, txOpts)
-		if err != nil && expSuccess {
-			t.Fatalf("%s: failed redeem transaction status: %v", test.name, err)
-		} else if receipt.GasUsed > expGas {
-			t.Fatalf("%s: gas used %d is > expected max %d", test.name, receipt.GasUsed, expGas)
+		receipt, err = test.redeemerClient.transactionReceipt(ctx, tx.Hash())
+		if err != nil {
+			t.Fatalf("%s: failed to get redeem receipt: %v", test.name, err)
 		}
 		spew.Dump(receipt)
+
+		expSuccess := !test.expectRedeemErr && test.addAmt
+		err = checkTxStatus(receipt, txOpts.GasLimit)
+		if err != nil && expSuccess {
+			t.Fatalf("%s: failed redeem transaction status: %v", test.name, err)
+		}
 		fmt.Printf("Gas used for %d redeems, success = %t: %d \n", len(test.swaps), expSuccess, receipt.GasUsed)
 
 		bal, err := balance()
@@ -1436,11 +1452,17 @@ func testRefund(t *testing.T, assetID uint32) {
 			t.Fatalf("%s: post-refund mining error: %v", test.name, err)
 		}
 
-		receipt, err := test.refunderClient.checkTxStatus(ctx, tx, txOpts)
-		if err != nil && test.addAmt {
-			t.Fatalf("%s: failed redeem transaction status: %v", test.name, err)
+		receipt, err := test.refunderClient.transactionReceipt(ctx, tx.Hash())
+		if err != nil {
+			t.Fatalf("%s: failed to get refund receipt: %v", test.name, err)
 		}
 		spew.Dump(receipt)
+
+		err = checkTxStatus(receipt, txOpts.GasLimit)
+		// test.addAmt being true indicates the refund shoud succeed.
+		if err != nil && test.addAmt {
+			t.Fatalf("%s: failed refund transaction status: %v", test.name, err)
+		}
 		fmt.Printf("Gas used for refund, success = %t: %d \n", test.addAmt, receipt.GasUsed)
 
 		// Balance should increase or decrease by a certain amount

